internal/dockercompose/client.go

```diff
@@ -44,34 +44,46 @@ func NewDockerComposeClient(env docker.LocalEnv) DockerComposeClient {
 }
 
 func (c *cmdDCClient) Up(ctx context.Context, configPaths []string, serviceName model.TargetName, shouldBuild bool, stdout, stderr io.Writer) error {
-	// docker-compose up is not thread-safe, because network operations are non-atomic. See:
-	// https://github.com/tilt-dev/tilt/issues/2817
-	c.mu.Lock()
-	defer c.mu.Unlock()
-
-	var args []string
+	var genArgs []string
 	if logger.Get(ctx).Level().ShouldDisplay(logger.VerboseLvl) {
-		args = []string{"--verbose"}
+		genArgs = []string{"--verbose"}
 	}
 
 	for _, config := range configPaths {
-		args = append(args, "-f", config)
+		genArgs = append(genArgs, "-f", config)
 	}
 
-	args = append(args, "up", "--no-deps", "-d")
-
 	if shouldBuild {
-		args = append(args, "--build")
-	} else {
+		var buildArgs = append([]string{}, genArgs...)
+		buildArgs = append(buildArgs, "build", serviceName.String())
+		cmd := c.dcCommand(ctx, buildArgs)
+		cmd.Stdout = stdout
+		cmd.Stderr = stderr
+		err := cmd.Run()
+		if err != nil {
+			return FormatError(cmd, nil, err)
+		}
+	}
+
+	// docker-compose up is not thread-safe, because network operations are non-atomic. See:
+	// https://github.com/tilt-dev/tilt/issues/2817
+	// TODO(nick): It might make sense to use a CondVar so that we can log a message
```

Comment:
(maybe worth adding a note that `docker-compose build` (as called above) CAN run in parallel, so we don't need to lock the mutex for it?)

Reply:
done